07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
 		}
 
-		// Ideally don't want to have to 
-		SPhysicalDeviceFeatures getPreferredDeviceFeatures() const override
+		// CREATE AND MAP HISTOGRAM BUFFER
 		{
-			auto retval = device_base_t::getPreferredDeviceFeatures();
-			retval.shaderStorageImageReadWithoutFormat = true;
-			return retval;
+			IGPUBuffer::SCreationParams gpuBufCreationParams;
+			gpuBufCreationParams.size = COMBINED_HISTOGRAM_BUFFER_BYTE_SIZE;
+			gpuBufCreationParams.usage = IGPUBuffer::E_USAGE_FLAGS::EUF_TRANSFER_DST_BIT | IGPUBuffer::E_USAGE_FLAGS::EUF_TRANSFER_SRC_BIT | IGPUBuffer::E_USAGE_FLAGS::EUF_STORAGE_BUFFER_BIT;
+			histogramBuffer = m_device->createBuffer(std::move(gpuBufCreationParams));
+			if (!histogramBuffer)
+				logFailAndTerminate("Failed to create a GPU Buffer of size %d!\n", COMBINED_HISTOGRAM_BUFFER_BYTE_SIZE);
+
+			histogramBuffer->setObjectDebugName("Histogram Buffer");
+
+			nbl::video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = histogramBuffer->getMemoryReqs();
+			// you can simply constrain the memory requirements by AND-ing the type bits of the host visible memory types
+			reqs.memoryTypeBits &= m_physicalDevice->getHostVisibleMemoryTypeBits() & m_physicalDevice->getDeviceLocalMemoryTypeBits();
+
+			m_histogramBufferAllocation = m_device->allocate(reqs, histogramBuffer.get(), nbl::video::IDeviceMemoryAllocation::E_MEMORY_ALLOCATE_FLAGS::EMAF_NONE);
+			if (!m_histogramBufferAllocation.isValid())
+				logFailAndTerminate("Failed to allocate Device Memory compatible with our GPU Buffer!\n");
+			assert(histogramBuffer->getBoundMemory().memory == m_histogramBufferAllocation.memory.get());
+
+			auto memoryRange = IDeviceMemoryAllocation::MemoryRange(0, m_histogramBufferAllocation.memory->getAllocationSize());
+
+			m_histogramBufferMemoryRanges[0] = ILogicalDevice::MappedMemoryRange(histogramBuffer->getBoundMemory().memory, 0, HISTOGRAM_BYTE_SIZE);
+			m_histogramBufferMemoryRanges[1] = ILogicalDevice::MappedMemoryRange(histogramBuffer->getBoundMemory().memory, HISTOGRAM_BYTE_SIZE, HISTOGRAM_BYTE_SIZE);
+			m_histogramBufferMemoryRanges[2] = ILogicalDevice::MappedMemoryRange(histogramBuffer->getBoundMemory().memory, HISTOGRAM_BYTE_SIZE * 2, HISTOGRAM_BYTE_SIZE);
+
+			m_histogramBufferMemPtrs[0] = static_cast<uint32_t*>(m_histogramBufferAllocation.memory->map(memoryRange, IDeviceMemoryAllocation::EMCAF_READ_AND_WRITE));
+			if (!m_histogramBufferMemPtrs[0])
+				logFailAndTerminate("Failed to map the Device Memory!\n");
+			m_histogramBufferMemPtrs[1] = m_histogramBufferMemPtrs[0] + HISTOGRAM_SIZE;
+			m_histogramBufferMemPtrs[2] = m_histogramBufferMemPtrs[1] + HISTOGRAM_SIZE;
 		}
 
-};
+		// PROCESS IMAGES
+		for (uint32_t imageToProcessId = 0; imageToProcessId < IMAGE_CNT; imageToProcessId++)
+		{
+			// TODO: wait for atomic instead? will it be even valid to do? we need to wait for gpu to finish anyway..
+			waitForPreviousStep(m_imagesLoadedSemaphore.get(), imageToProcessId + 1);
```

Comment:
atomic would have been LoadThread->ProcessThread

what you have now is LoadThread->UploadQueue->ProcessThread->ProcessQueue

What you want is LoadThread->UploadQueue->ProcessQueue


you need a `m_imagesLoadedSemaphore, imageToProcessId+1, COPY_BIT` wait in the `submitInfo.waitSemaphores`

Reply:
but here: https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/vulkan_1_3_example_07_fix/07_StagingAndMultipleQueues/main.cpp#L419-L426 i'm using a GPU image in CPU to update descriptor set write, so i'm touching the loaded images with CPU no?

Reply:
no, you're touching the Descriptor set with CPU